ens-improvement-proposals/ensip-15-normalization-standard.md

```diff
@@ -0,0 +1,343 @@
+* For user convenience, leading and trailing whitespace should be trimmed before normalization, as all whitespace codepoints are disallowed.  Inner characters should remain unmodified.
+* No string transformations (like case-folding) should be applied.
+
+1. Split the name into [labels](./ensip-1-ens#name-syntax).
+1. [Normalize](#normalize) each label.
+1. Join the labels together into a name again. 
+1. The result is normalized and ready for [hashing](./ensip-1-ens#namehash-algorithm).
+
+### Normalize
+
+1. [Tokenize](#tokenize) — transform the label into **Text** and **Emoji** tokens.
+	* If there are no tokens, the label cannot be normalized.
+1. Apply NFC to each **Text** token.
+1. Strip `FE0F` from each **Emoji** token.
+1. [Validate](#validate) — check if the tokens are valid and obtain the **Label Type**.
+	* The **Label Type** and **Restricted** state may be presented to user for additional security.
+1. Concatenate the tokens together.
+	* Return the normalized label.
+
+### Tokenize
+
+Given a string, convert to codepoints, and produce a list of **Text** and **Emoji** tokens, each with a payload of codepoints.
+
+1. Allocate an empty codepoint buffer.
+1. Find the longest emoji sequence that matches the remaining input.
+	* `FE0F` is optional from the input during matching.
+	* Example: `Emoji[A FE0F B]` matches `[A FE0F B]` and `[A B]` but not `[A FE0F FE0F B]`
+1. If an emoji sequence is found:
+	* If the buffer is nonempty, emit a **Text** token, and clear the buffer.
+	* Emit an **Emoji** token with the fully-qualified matching sequence.
+	* Remove the matched sequence from the input.
+1. Otherwise:
+	1. Remove the leading codepoint from the input.
+	1. Determine the codepoint type:
+		* If **valid**, append the codepoint to the buffer.
+			* This set can be precomputed from the union of characters in all groups and their NFD decompositions.
+		* If **mapped**, append the corresponding mapped codepoint(s) to the buffer.
+		* If **ignored**, do nothing.
+		* Otherwise, the label cannot be normalized.
+1. Repeat until all the input is consumed.
+1. If the buffer is nonempty, emit a final **Text** token.
+
+### Validate
+
+Given a list of **Emoji** and **Text** tokens, determine if the composition is valid and return the **Label Type**.
+
+1. If only **Emoji** tokens:
+	* Return `"Emoji"`
+1. If a single **Text** token and every characters is ASCII (`00..7F`):
+	* `5F (_) LOW LINE` can only occur at the start.
+		* Must match `/^_*[^_]*$/`
+	* The 3rd and 4th characters must not both be `2D (-) HYPHEN-MINUS`.
+		* Must not match `/^..--/`
+	* Return `"ASCII"`
+		* The label is free of **Fenced** and **Combining Mark** characters, and not confusable.
+1. Concatenate all the tokens together.
+	* `5F (_) LOW LINE` can only occur at the start.
+		* Must match `/^_*[^_]$/u`
+	* The first and last characters cannot be **Fenced**.
+	* **Fenced** characters cannot be contiguous.

```

Comment:
Yes.  It would be invalid.

I added the following:

... If any assertion fails, the name cannot be normalized.

* The first and last characters cannot be **Fenced**.
    * Example: `"a’s"` and `"a・a"` are valid, `"’85"` and `"joneses’"` and `"・a・"` are invalid.
* **Fenced** characters cannot be contiguous.
    * Example: `"a・a’s"` is valid, `"6’0’’"` and `"a・・a"` are invalid.